### PR TITLE
Backend/get other profile

### PR DESCRIPTION
**Pull request by notiona**
## Description

This PR implements GET /api/profile/:user_id/, which gets profile of another user, if the current user has read permission (is onechon or twochon)

Test coverage is not 100%, as many changes and refactoring is concurrently taking place.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: backend/linklink/views.py [hunk 1]
```diff
@@ -522,6 +522,95 @@ def profile(request):
...
                     "message":
-                    "Profile not found. Your profile is not created yet."
+                    f"Profile userId={request.user.linklinkuser.id} not found."
+                }
+            )
+
+
+@allowed_method_or_405(["GET"])
+@logged_in_or_401
+def other_profile(request, user_id):
```
Comment (hunkim98) [inline backend/linklink/views.py:532]: 굳이 my_profile과 other_profile을 구분할 필요가 있나요? 어차피 id로 profile정보를 조회하는 건데 my_profile이랑 other_profile로 나눈 이유가 궁금합니다

**Review by hunkim98 — COMMENTED** [review]
고생하셨습니다. 코멘트 확인부탁드립니다

File: backend/linklink/views.py (hunk 1 above)
Comment (notiona) [reply backend/linklink/views.py:532]: my_profile은 나의 Profile에 대한 API(GET, PUT)이고, other_profile은 다른 사람(1촌 또는 2촌)의 Profile에 대한 API(GET)입니다.

Comment (notiona) [reply backend/linklink/views.py:532]: 현재는 my_profile은 GET, POST, PUT request를 받지만 다음 fix&refactor branch에서 아까 논의한대로 GET과 PUT만 받을 예정입니다.

Comment (hunkim98) [reply backend/linklink/views.py:532]: 알겠습니다. my_profile에서 GET은 아마 프로필 수정 페이지에서 본인의 profile정보를 조회할 때만 쓰이고 쓰지는 않을 것 같네요.(my_profile에서의 GET을 없애고 other_profile GET으로 프로필 정보 조회를 다 통틀어 해결해도 좋다고 생각하지만 실제 구현은 원하시는대로 하시면 되겠습니다). 

other_profile GET은 frontend쪽에서 linklink.com/profile/:userId 로 불릴 것으로 보이는데 유저가 userId param에 자신의 아이디를 넣으면 결국에는 자신의 profile을 보는 것과 마찬가지라서 other_profile 이라는 이름이 헷갈렸습니다.

Comment (notiona) [reply backend/linklink/views.py:532]: my_profile의 GET과 other_profile의 GET은 url의 차이도 있지만, 결정적으로 Quality Tag를 response body에 포함시키는지 여부도 있기에 별도의 API로 구현했습니다.
그렇기에 현재 구현으로 유저가 자신의 user_id로 other_profile을 부르면 QualityTag도 불러올 수 있기에 이 경우는 Forbidden을 반환하게 구현되어 있습니다.

Comment (notiona) [reply backend/linklink/views.py:532]: "만약 404가 뜨지 않게하려면 frontend쪽에서 profile/:userId 에서 userId param에 따라 currentUserId와 일치하는 여부에 따라 axios get을 다르게 호출해야 하는데 그렇게 할 경우 프론트쪽에서는 필요없는 복잡한 로직이 생길 것 같네요."

이게 그렇게 복잡한 로직인가요?

말씀하신대로 other_profile에서 모든 GET을 처리하는 로직으로 백엔드 API를 수정한다고 해도 `axios.get('/api/user/profile/:currentUserId/')`를 호출하기 위해 currentUserId를 알아야 해서 `axios.get('/api/user/profile/')`이라고 불러도 구현상의 어려움은 더 복잡해질 것 같지 않습니다.

오히려 백엔드에서 현재 other_profile에서 GET을 할 수 있는 permission 제한을 (onechon OR twochon)으로 제한하고 있는데, 이걸 (currentUser OR onechon OR twochon)으로 늘리고, SkillTag 반환여부를 분기문으로 구현하는게 더 불필요한 로직일 것 같습니다.

Comment (notiona) [reply backend/linklink/views.py:532]: + 추가로 현재 구현상에서는 유저가 자신의 currentUserId로 `axios.get('/api/user/profile/:currentUserId/')`를 호출하면 404가 아닌 403({"message": "No read permission for userId={userId}."})이 뜹니다. 이 장치가 자신의 QualityTag를 보려는 유저의 접근을 제한하는 효과도 고려했습니다.

Comment (hunkim98) [reply backend/linklink/views.py:532]: 여기서 말하는 복잡한 로직이란 구현이 어렵다는 얘기가 아닙니다. 여기서 제가 말하는 복잡함이란, 불필요한 분기문이 추가되어 생기는 긴 코드라고 보시면 되겠습니다.

## 문제1. 프론트에서 불필요한 코드로직이 추가된다

직접 react 코드를 짜드려서 보여드리겠습니다

### /api/profile 로 현재 유저의 프로필을 조회하고 /api/profile/:otherUserId 로 다른 사람 것 불러올 시

```ts
// Profilt.tsx

const Profile = () => {

const {otherUserId} = useParams()

const dispatch = useAppDispatch()

const currentUser = useAppSelector((state) => state.users.currentUser));

useEffect(() => {
  if(otherUserId) {
    if(currentUser) {
      dispatch(getCurrentUserProfile())
    } else {
      dispatch(getOtherUserProfile(otherUserId))
    }
  }
}, [currentUser, otherUserId])

//... other logic
}
```

이렇게 됩니다.

만약 하나의 get으로 모든 profile get을 처리하면 어떻게 될까요?

```ts
// Profilt.tsx

const Profile = () => {

const {userId} = useParams()

const dispatch = useAppDispatch()

const currentUser = useAppSelector((state) => state.users.currentUser));

useEffect(() => {
  if(otherUserId) {
      dispatch(getProfile(userId))
  }
}, [userId])

//... other logic
}
```

훨씬 짧습니다. 또한 의문이 들지 않는 코드죠

(더 진지하게 들어가면) 지금 getCurrentProfile와 getOtherUserProfile로 나눌 경우 굳이 이후 다른 개발자가 이 코드를 볼 시 왜 분기를 했는지 파고들어가야 합니다.

Comment (hunkim98) [reply backend/linklink/views.py:532]: ## 문제2. 백에서 불필요한 로직이 추가되며 이후 수정이 필요할 경우 수정이 힘들어진다

사실 프론트에서의 복잡함은 그다지 중요하지 않을 수 있습니다. 백에서 주어지는 대로 그냥 하면 그만이죠. 더 큰 문제는 사실 백에 있습니다. 

Screen Shot 2022-11-16 at 1 11 44 PM

위 이미지는 현재 작업하고 계신 브랜치의 views.py를 가져온 사진입니다. 이미지를 보시면 왼쪽은 my_profile, other_profile사진입니다.  response_dict는 qualityTags를 제외하고는 로직이 똑같습니다. 유사 duplicate이죠. my_profile get이랑 other_profile get이 상당한 차이가 있을 경우 나누는 것을 추천합니다. 하지만 지금의 경우 response_dict의 json key value하나 빼고는 다른 게 없습니다. 그러면 과연 필요한 나눔인지 잘 모르겠습니다.

duplicate문제보다 더 심각한 문제가 있습니다. Worst Case Scenario를 가정해보죠. 저희가 만약 **get profile의 response를 수정할 일이 있다고 합시다. profile에 새로운 field가 추가되어서 response_dict에 json key를 하나 추가해야 한다고 합시다.** 그럴 경우 어떻게 될까요? my_profile get이랑 other_profile get 둘다를 수정할 필요가 있죠? 수정사항이 작으면 문제 없습니다. 하지만 수정사항이 잦아지면 어떡할까요? 처음 코드를 짠 사람 이외에는 어디가 수정이 필요한지 모를 수 있습니다. 최악의 경우에는 other_profile get만 업데이트하고 my_profile get은 업데이트하는 것을 깜빡할 수 있지요.

Comment (hunkim98) [reply backend/linklink/views.py:532]: ## 문제3. 실제 사례를 통해서 배우자 (instagram)

그럼 한번 실제 회사들은 profile을 어떻게 관리하는지 한번 보는게 좋겠습니다. 다른 기업들은 my_profile이랑 other_profile을 부를때 다른 api_url로 부를까요? 아닙니다. 하나의 api_url로 관리를 합니다.

**instagram을 한번 보겠습니다**

### my profile

Screen Shot 2022-11-16 at 1 19 59 PM


### other person pr

Screen Shot 2022-11-16 at 1 21 34 PM 1
ofile

위 사진은 제 프로필을 확인할 때 불리는 url이고 아래 사진은 남의 프로필을 확인할 때 불리는 url입니다. 차이가 있나요? 없습니다. username query만 달라지고 실제 call하는 url은 같은 것을 확인할 수 있습니다. 왜 이렇게 했을까요? profile get하는 것은 내 것을 보거나 남의 것을 보거나 **front에서 render하는 것이 같기 때문입니다** 

사실 profile get을 하나로 다 통일하면 좋은 점이 많습니다. 
1. api route의 모호성이 사라진다 (GET profile 하나만 있으면 개발자는 이 api url이 무슨 역할을 하는지 큰 생각 없이 이해할 수 있습니다)
2. test code를 짜기가 쉬워진다. (같은 역할을 하는 함수를 위해 두번 테스트 코드를 작성할 필요가 없어지죠)

Comment (dongjoocha) [reply backend/linklink/views.py:532]: 저도 의견 보태자면 정훈 님께서 `my_profile`과 `other_profile`을 구분하신 동기는 충분히 이해가 갑니다만, 동훈 님께서 언급하셨듯이 1) response_dict의 json key value 하나를 제외하고는 코드가 다른 점이 크게 없다는 점

Comment (hunkim98) [reply backend/linklink/views.py:532]: ### 정리 및 제안

그래도 의문이 들 수 있습니다. "아니 내 프로필을 qualityTags key를 빼고 상대방 프로필은 qualityTags key를 넣어야 한다고!" 자 그럼 문제 발생지인 response를 볼까요? 

지금 my profile get과 other profile get을 나눈 이유는 qualityTags하나입니다. 그럼 굳이 나누는 것만이 유일한 방법일까요? 아닙니다. response json의 차이 때문에 굳이 분기했다면 같은 response를 보내되 request.user.id가 있을 경우 빈 qualityTags Array를 보내면 되지 않을까요? json response로 qualityTags를 확인할 수 있지만 빈 array면 사실상 정보가 없는 거랑 마찬가지이지요. 그렇게 빈 qualityTags array가 오면 애초에 front는 qualityTags를 위한 컴포넌트를 렌더하지 못하지요.

Comment (notiona) [reply backend/linklink/views.py:532]: 자세한 설명 감사합니다. 시간을 많이 투자해 더 좋은 best practice가 무엇인지 다 같이 고민하는 모습이 아주 좋네요. 말씀하신 front의 로직이 간결해지는 것과 현재 백엔드 코드의 문제점, Intagram등의 웹사이트에서 진행하는 best practice 내용 잘 이해했습니다. 저 스스로는 아직 이것이 우리 프로젝트를 위한 best practice이고 이 결정이 앞으로도 바뀌지 않는다는 보장을 할 수 없으나, 다른 분들께서 그렇게 생각하시고 저 또한 두 방식의 차이가 크지 않다고 생각해 제안 주신 방향으로 백엔드 코드를 수정해보도록 하겠습니다. 이 수정사항은 PR #76 에 추가해 반영하도록 하겠습니다.

 다만 수정에 앞서 말씀해주신 내용에 대한 제 생각을 적어보겠습니다. 앞으로 백엔드 코드의 refactoring 방향성과 구현된 백엔드에 대한 제 개인적인 생각이니 가볍게 읽어주시면 감사하겠습니다.

 우선 현재 백엔드 코드에서 일어나는 duplication 문제에 대해 언급이 많으셔서 이 점에 대해서 말씀을 드리겠습니다. 말씀하신대로 views.py의 profile 관련 API에서, 특히 response body를 패키징하는 작업에서 duplication이 발생합니다. 이는 최대한 빨리 profile view를 구현하는 과정에서 나온 것으로, 추후 helper 함수 (package_profile_response)를 통해 refactoring할 예정입니다. 다만 이러한 문제는 다른 view를 보면 아실 수 있지만 profile view 만의 문제는 아닙니다. Auth와 Get Friend의 API에서도 현재 view 코드에서는 response body를 packaging할때 duplication이 발생하고 있습니다. 그러므로 "백엔드 profile 코드에서 duplication이 발생하기에 GET을 두개로 쪼개는 것은 좋지 않다"의 주장은 refactoring 이후에는 발생하지 않을 문제이고, profile view에서만 나타나는 문제도 아닙니다. 또한 test.py의 다양한 Object Initialization에서도 현재 중복이 많이 발생하고 있어 추후 views.py의 response body packaging과 test.py의 object initialization은 helper 함수로 refactoring을 염두해두고 있다는 점을 알아주셨으면 합니다.

한가지 더 말씀드리고 싶은 점은, 프론트를 구현하시면서 어떤 방향으로 백엔드 API의 동작을 기대하고 그것을 docs로 합의했다면, 이후에는 최대한 이 방향성을 바꾸지 않았으면 하는 바람입니다. 현재 백엔드 코드는 test coverage 100%를 달성함과 동시에 edge case를 잡기 위해 이미 테스트된 함수에 대해서도 여러 테스트를 진행하고 있습니다. 그렇기에 한번 합의된 구현 이후로 이미 구현된 백엔드의 schema나 API의 동작 형태를 조금이라도 바꾼다면, views 코드가 바뀌는 것의 몇배로 docs와 test 코드를 수정해야 하기에, 작업에 더욱 시간이 많이 소요되고 있습니다.

다시 한번 좋은 논의 감사드립니다!

Comment (dongjoocha) [reply backend/linklink/views.py:532]: duplication 관련해서 첨언하자면 front, back 구분 없이 test.py와 같은 **test 목적의 코드**의 경우 반복적인 object initialization이 필수적이고 helper 함수로 refactoring을 수행해도 불가피한 duplication이 발생할 수밖에 없다고 생각합니다. 그래서 제가 sonarcloud configuration을 수정할 때 test 코드 자체에 대해서는 coverage, duplication rate 등 code quality 분석을 하지 않도록 제외해두었고, 추후에 이와 같이 제외할 파일이 추가되면 언제든 추가할 수 있으니 참고하시면 좋을 것 같습니다!

docs로 합의한 부분에 대해 수정이 필요한 경우 소요되는 노력과 시간이 만만치 않다는 점을 remind 해주신 점 감사드립니다. docs 작성 시 최대한 구체적으로 논의를 해서 앞으로는 docs에 명시된 방향을 변경하는 경우를 최소화 해야겠습니다. 그럼에도 어쩔 수 없이 추후 docs와 다른 방향으로 변경이 필요한 경우가 생길 수도 있는데, 이때에는 front측에서도 그 변화에 소요되는 노력과 시간을 감안하고서라도 꼭 필요할 만큼 중요한 사안임을 알아주시면 감사할 것 같습니다.

Comment (JadenJSPark) [reply backend/linklink/views.py:532]: 저번에도 말씀드렸지만 각자 최우선적으로 참고하는 문서와 소통 방식이 다른 것 같아서, 변경 사항이 결정되면 최대한 빨리 카톡 언급 기능을 통해 서로 알려주는 것이 최선일 것 같습니다.

Comment (notiona) [reply backend/linklink/views.py:532]: 좋은 말씀 감사합니다.

test 코드에서는 duplication이 발생할 수 밖에 없지만, 현재 제가 작성한 백엔드 test 코드의 경우 setup 함수에서 object를 초기화하는 범위에 따라 각 unittest에서 추가적으로 초기화해야 하는 범위가 늘어나고, 그 구분(각 unittest가 object를 만들어야 하는지 setup이 object를 만들어야 하는지)도 코드가 커짐에 따라 점점 모호해지고 있습니다. 이 문제를 해결하기 위해 추후 test 코드를 리팩토링할 필요성을 꾸준히 느끼고 있습니다. 이 점은 부족한 제 테스트 코드를 리팩토링 하면서 조금씩 해결해보겠습니다.

Docs 작성 시 최대한 구체적으로 논의를 해서 앞으로는 docs에 명시된 방향을 변경하는 경우를 최소화 해야한다는 점에 동의합니다. 어쩔 수 없이 추후 docs와 다른 방향으로 변경이 필요한 경우가 생길 수 있음도 충분히 이해하고 있습니다. 수정을 가볍게 생각해 그때그때 "이렇게 바꾸자"라고 하는 행위만 앞으로도 경계하면 좋을 것 같습니다.

Comment (hunkim98) [reply backend/linklink/views.py:532]: 모두 좋은 논의 및 좋은 말씀 감사드립니다.

docs에 명시된 방법과 다른 방향으로 코드를 짬으로 인해 따라 이미 짰던 코드나 테스트가 전체 바뀌는 것에 따르는 고통이 크기에 수정을 가볍게 생각하는 것을 경계하자는 것에 동의합니다. 그 과정이 절대 순탄치 않다는 것을 잘 압니다. 저희가 docs에 짠 것이 보고서 마감기한에 맞춰서 급하게 짠 것이다보니 세부 디테일에 대해서 다르게 생각하고 합의가 부족할 수밖에 없었다고 생각합니다. 백엔드 쪽에서는 아무래도 docs의 세부작업이 많이 부실했는데, 이를 감안하고도 코드를 열심히 작성해주신 것에 감사하고 있습니다. 

dupllication에 대한 코멘트는 가볍게 코멘트한 것 뿐 전체 코드에 대한 평가가 아니니 이에 불만을 제기한 것이 아님을 알아주셨으면 합니다. 프론트 쪽도 dupllication이 넘쳐납니다😅. best quality code를 강조할 생각은 별로 없습니다. 작동만 한다만 저는 대만족합니다. 다만 예시를 들면서 의견을 작성한 것은 프론트나 백 모두 더 쉽고 간결하게 코드 및 테스트코드 (사실 테스트 코드를 많이 줄일 수 있다고 생각해서 제시한 게 가장 큽니다)를 짤 수 있지 않을까 라는 생각에서 나온 것이니 그렇게 이해주시면 감사하겠습니다.

Docs에 대한 합의가 중요한데, 큰 api작업인데 미리 합의점을 찾아야 하는 중요성에 동감합니다. 저도 notion의 몇부분을 작업하면서 수정을 하고 있기는 한데 내용이 많다보니 하나하나 일일히 확인 못한점을 사과드립니다. 앞으로 프론트에서 미리 구현한 게 있으면 백에 문의를 해서 합의점을 찾아보는 식의 과정이 있으면 좋겠네요.

개인취향이지만, 가능만 하면 테스트 코드를 리팩토링하거나 처음부터 완벽히 채우는 것에 대한 부담을 덜 느끼셨으면 합니다. 테스트코드 coverage가 중요하지만, 100%가 수업의 목적이 아니기에 크게 신경을 안 쓰셔도 좋을 것 같습니다. 애초에 프론트의 경우 처음부터 라이브러리를 안 썼는데, 이는 라이브러리를 활용할 경우 프론트가 테스트 코드를 작성하기 어려워지기 때문이고, 최대한 덜 써서 테스트 코드를 빠르고 쉽게 짜기 위해서였습니다. 마감일에 맞춰 백에서 테스트 커버리지가 낮아도 프론트에서 커버리지를 100으로 맞추는 것은 쉽게끔 만들었기 때문에 너무 걱정하지 않으셔도 될 것 같습니다. 애초에 백을 혼자 담당하시는데 운영코드랑 테스트 코드 둘 다 한꺼번에 만들기에 테스트 코드만큼은 부담을 덜 느끼셨으면 합니다.

**Comment by notiona** [comment]
논의가 끝났고 이 PR의 커밋은 PR #76 에 반영 되어 main에 merge되었음으로 이 pull request는 close 하겠습니다.
